proof/crefine/RISCV64/Finalise_C.thy

```diff
@@ -35,102 +35,226 @@ declare if_split [split del]
 definition
   "option_map2 f m = option_map f \<circ> m"
 
-lemma head_end_ksReadyQueues_':
-  "\<lbrakk> (s, s') \<in> rf_sr; ksReadyQueues_asrt s; pspace_aligned' s; pspace_distinct' s;
-     d \<le> maxDomain; p \<le> maxPriority \<rbrakk>
-   \<Longrightarrow> head_C (index (ksReadyQueues_' (globals s')) (cready_queues_index_to_C d p)) = NULL
-       \<longleftrightarrow> end_C (index (ksReadyQueues_' (globals s')) (cready_queues_index_to_C d p)) = NULL"
-  apply (frule (2) rf_sr_ctcb_queue_relation[where d=d and p=p])
-  apply (clarsimp simp: ready_queue_relation_def ksReadyQueues_asrt_def)
-  apply (drule_tac x=d in spec)
-  apply (drule_tac x=p in spec)
+lemma thread_state_ptr_set_tcbQueued_tcb_fields:
```

Comment:
This starts a chain of lemmas which are all vcg spec rules, but don't have anything in their name to differentiate them as such. I think they really should. I also have a question: why do you need so many of these?

Reply:
While I'm here, do you know what the option_map2 doing in this file? It doesn't have a FIXME move on it either...

Reply:
I'm not sure what comment would be good for these "`tcb_field`" rules, but maybe I could say before the first of them "some Hoare rules to show that many functions do not alter fields of concern within functions which cancel IPC". Does that sound OK?

And I'm not sure about `option_map2` being in this file, either. Should I move it to where `option_map` is? That is in `Lib.thy`.

Reply:
I think what Raf means is that the lemmas should have something in the name that identifies them as vcg rules (like we have `_wp` for wp rules etc). They are not full spec rules, so definitely don't call them only `_spec`, but they could be `_vcg_tcb_fields`, or `_spec_tcb_fields`.

To answer Raf's question why there are so many of them: they are exactly the same as `wp` rules for a specific property -- you need one for each function. We have so far not needed to do that much in CRefine, because we always were either able to rearrange the C so that there is no need for `vcg` or we were able to write a `spec` rule without unfolding inner functions much, but neither of these works here because there is a much longer chain of inner functions here than on master. Proving one lemma per function avoids the vcg blow-up you get with unfolding.

Reply:
`option_map2`: I'm not sure either. It would probably make sense to put it into Lib, but it is also unrelated to this PR, it's just using what was already there, we only see more of it because we now need to prove more lemmas that have it in their statement.

If we do clean it up and move to Lib, it should probably happen in its own commit so we can cherry-pick that onto master.

Reply:
Yes, that's what I meant with the names. I use these rarely, so don't know what naming convention we want, but they are alternative spec rules, so I'd go with `spec` unless there's a pressing need to do otherwise.

Yeah, for `option_map2` definitely not in this PR, I only noticed it by coincidence and it looks very silly where it is.